### ISSUE TITLE
Faster joins: Update room stats once state re-sync completes

### ISSUE THREAD

**Issue by squahtx** [matrix-org/synapse#12814]
---

Potentially useful / possible blockers:

- [x] #14419 — might want to track the stream of un-partial-stated rooms and do a recomputation when it completes

**Comment by richvdh** [comment]
Also: what do the room stats say *during* re-sync, and is that ok?

**Comment by richvdh** [comment]
https://github.com/matrix-org/synapse/blob/7c6b2204d143550d81e5bf9612c4e69fe0866b4c/synapse/handlers/federation.py#L1561-L1570

**Comment by H-Shay** [comment]
Looking at the code, it seems that there are two tables that would need updating: `room_stats_state`, and `room_stats_current`. Is there any others that I am missing?
As for an approach to this, would it make sense to just pull the current state map and use the info contained within to update the tables? It appears that most of the already-existing functions to update the stats tables rely on using state deltas, which seems less useful in this context?

**Comment by squahtx** [comment]
> Looking at the code, it seems that there are two tables that would need updating: room_stats_state, and room_stats_current. Is there any others that I am missing?

Those are all the tables for this issue I believe.

As for the approach:

**Non-faster joins operation**
Normally, when persisting an event, we
1. [call](https://github.com/matrix-org/synapse/blob/54c012c5a8722725cf104fa6205f253b5b9b0192/synapse/storage/databases/main/events.py#L485) [`PersistEventsStore._update_current_state_txn`](https://github.com/matrix-org/synapse/blob/54c012c5a8722725cf104fa6205f253b5b9b0192/synapse/storage/databases/main/events.py#L1046) with the state delta due to the event.
2. `PersistEventsStore._update_current_state_txn` inserts the delta as rows into `current_state_delta_stream` under a given, new, stream id (created using `store._stream_id_gen`).
3. I believe we call `Notifier.notify_replication()` somewhere, but I have not looked.

`Notifier.notify_replication()` runs 4 callbacks:
* [`PresenceHandler.notify_new_event`](https://github.com/matrix-org/synapse/blob/54c012c5a8722725cf104fa6205f253b5b9b0192/synapse/handlers/presence.py#L1319)
* [`StatsHandler.notify_new_event`](https://github.com/matrix-org/synapse/blob/54c012c5a8722725cf104fa6205f253b5b9b0192/synapse/handlers/stats.py#L65)
* [`UserDirectoryHandler.notify_new_event`](https://github.com/matrix-org/synapse/blob/54c012c5a8722725cf104fa6205f253b5b9b0192/synapse/handlers/user_directory.py#L110)
* [`ReplicationStreamer.on_notifier_poke`](https://github.com/matrix-org/synapse/blob/54c012c5a8722725cf104fa6205f253b5b9b0192/synapse/replication/tcp/resource.py#L110)

`StatsHandler.notify_new_event` starts off a processing loop which consumes new `current_state_delta_stream` rows and updates `room_stats_state` and `room_stats_current`.

**Faster joins operation**
When un-partial stating a room, we [call](https://github.com/matrix-org/synapse/blob/54c012c5a8722725cf104fa6205f253b5b9b0192/synapse/handlers/federation.py#L1723) [`StateHandler.update_current_state`](https://github.com/matrix-org/synapse/blob/54c012c5a8722725cf104fa6205f253b5b9b0192/synapse/state/__init__.py#L560), which also eventually calls `PersistEventsStore._update_current_state_txn` with the state delta due to un-partial stating the room and a new stream id (created using `store._stream_id_gen`). [We then call `Notifier.notify_replication()`](https://github.com/matrix-org/synapse/blob/54c012c5a8722725cf104fa6205f253b5b9b0192/synapse/handlers/federation.py#L1735-L1737).

So in single process mode, it looks like things might already just work? We'll want to write a complement test to confirm it.

**Worker mode**
We need to convince ourselves that `Notifier.notify_replication` gets called on other workers appropriately, otherwise room stats will lag behind under certain circumstances (when the stats processing loop isn't running until we receive another, possibly unrelated, event).

I believe `StateHandler.update_current_state` causes us to send `EventsStreamCurrentStateRow`s over replication.
[We currently don't trigger `Notifier.notify_replication` in response, which will need fixing.](https://github.com/matrix-org/synapse/blob/54c012c5a8722725cf104fa6205f253b5b9b0192/synapse/replication/tcp/client.py#L136)

---

In summary, we might expect things to just work in single process mode. We'll want to write a complement test for it.
In worker mode, we are missing a `notify_replication` call somewhere and we would expect the complement test to flake until it's added.

Hopefully the above isn't too confusing or unclear.

**Comment by H-Shay** [comment]
That's extremely clear, thank you very much!

**Comment by squahtx** [comment]
> Also: what do the room stats say _during_ re-sync, and is that ok?

During re-sync, the room stats reflect the current state of the room and exclude all remote users except new joiners.
